File: ios/SafeBox/Notes/MarkdownPreview.swift
import SwiftUI

/// Renders the shared cross-platform markdown subset (idea plan §3.2):
/// headings, bold/italic, inline code, bullet & numbered lists, and checklists
/// as non-interactive styled text. Line-aware block pre-pass over
/// AttributedString(markdown:) inline rendering.
struct MarkdownPreview: View {
    let markdown: String

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 6) {
                ForEach(Array(markdown.components(separatedBy: .newlines).enumerated()), id: \.offset) { _, line in
                    lineView(line)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding()
        }
    }

    @ViewBuilder
    private func lineView(_ rawLine: String) -> some View {
        let line = rawLine.trimmingCharacters(in: .whitespaces)
        if line.isEmpty {
            Spacer().frame(height: 4)
        } else if line.hasPrefix("# ") {
            inlineText(String(line.dropFirst(2))).font(.title.bold())
        } else if line.hasPrefix("## ") {
            inlineText(String(line.dropFirst(3))).font(.title2.bold())
        } else if line.hasPrefix("### ") {
            inlineText(String(line.dropFirst(4))).font(.title3.bold())
        } else if let checked = checklistContent(line, checked: true) {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Image(systemName: "checkmark.square")
                inlineText(checked).strikethrough(false)
            }
        } else if let unchecked = checklistContent(line, checked: false) {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Image(systemName: "square")
                inlineText(unchecked)
            }
        } else if line.hasPrefix("- ") || line.hasPrefix("* ") || line.hasPrefix("+ ") {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Text("•")
                inlineText(String(line.dropFirst(2)))
            }
        } else if let (number, content) = numberedItem(line) {
            HStack(alignment: .firstTextBaseline, spacing: 6) {
                Text("\(number).")
                inlineText(content)
            }
        } else {
            inlineText(line)
        }
    }

    private func checklistContent(_ line: String, checked: Bool) -> String? {
        let markers = checked ? ["- [x] ", "- [X] ", "* [x] ", "* [X] "] : ["- [ ] ", "* [ ] "]
        for marker in markers where line.hasPrefix(marker) {
            return String(line.dropFirst(marker.count))
        }
        return nil
    }

    private func numberedItem(_ line: String) -> (String, String)? {
        guard let dotIndex = line.firstIndex(of: "."),
              !line[line.startIndex..<dotIndex].isEmpty,
              line[line.startIndex..<dotIndex].allSatisfy({ $0.isNumber }),
              line.index(after: dotIndex) < line.endIndex,
              line[line.index(after: dotIndex)] == " " else { return nil }
        return (String(line[line.startIndex..<dotIndex]),
                String(line[line.index(dotIndex, offsetBy: 2)...]))
    }

    private func inlineText(_ content: String) -> Text {
        if let attributed = try? AttributedString(
            markdown: content,
            options: AttributedString.MarkdownParsingOptions(interpretedSyntax: .inlineOnlyPreservingWhitespace)
        ) {
            return Text(attributed)
        }
        return Text(content)
    }
}

File: android/app/src/main/java/com/calcplus/calculator/feature/notes/MarkdownPreview.kt
package com.calcplus.calculator.feature.notes

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.withStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp

/**
 * Renders the shared cross-platform markdown subset (idea plan §3.2):
 * headings, bold/italic, inline code, bullet & numbered lists, and checklists
 * as non-interactive styled text. Line-aware block pre-pass over an inline
 * AnnotatedString renderer — the sanctioned custom fallback (android-plan
 * §4.3), mirroring the iOS MarkdownPreview line by line.
 */
@Composable
fun MarkdownPreview(markdown: String, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        markdown.lines().forEach { rawLine ->
            MarkdownLine(rawLine.trim())
        }
    }
}

@Composable
private fun MarkdownLine(line: String) {
    when {
        line.isEmpty() -> Spacer(modifier = Modifier.height(4.dp))
        line.startsWith("# ") -> Text(
            inlineMarkdown(line.removePrefix("# ")),
            style = MaterialTheme.typography.headlineMedium,
            fontWeight = FontWeight.Bold,
        )
        line.startsWith("## ") -> Text(
            inlineMarkdown(line.removePrefix("## ")),
            style = MaterialTheme.typography.headlineSmall,
            fontWeight = FontWeight.Bold,
        )
        line.startsWith("### ") -> Text(
            inlineMarkdown(line.removePrefix("### ")),
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold,
        )
        checklistContent(line, checked = true) != null -> ChecklistRow("☑", checklistContent(line, true)!!)
        checklistContent(line, checked = false) != null -> ChecklistRow("☐", checklistContent(line, false)!!)
        line.startsWith("- ") || line.startsWith("* ") || line.startsWith("+ ") ->
            ChecklistRow("•", line.drop(2))
        numberedItem(line) != null -> {
            val (number, content) = numberedItem(line)!!
            ChecklistRow("$number.", content)
        }
        else -> Text(inlineMarkdown(line), style = MaterialTheme.typography.bodyLarge)
    }
}

@Composable
private fun ChecklistRow(marker: String, content: String) {
    Row(verticalAlignment = Alignment.Top) {
        Text(marker, style = MaterialTheme.typography.bodyLarge)
        Text(
            inlineMarkdown(content),
            style = MaterialTheme.typography.bodyLarge,
            modifier = Modifier.padding(start = 8.dp),
        )
    }
}

private fun checklistContent(line: String, checked: Boolean): String? {
    val markers = if (checked) {
        listOf("- [x] ", "- [X] ", "* [x] ", "* [X] ")
    } else {
        listOf("- [ ] ", "* [ ] ")
    }
    for (marker in markers) {
        if (line.startsWith(marker)) return line.removePrefix(marker)
    }
    return null
}

private fun numberedItem(line: String): Pair<String, String>? {
    val dotIndex = line.indexOf('.')
    if (dotIndex > 0 &&
        line.substring(0, dotIndex).all { it.isDigit() } &&
        dotIndex + 1 < line.length && line[dotIndex + 1] == ' '
    ) {
        return line.substring(0, dotIndex) to line.substring(dotIndex + 2)
    }
    return null
}

/** Inline subset: **bold**, *italic*, `code`. */
internal fun inlineMarkdown(text: String): AnnotatedString = buildAnnotatedString {
    var i = 0
    while (i < text.length) {
        when {
            text.startsWith("**", i) -> {
                val end = text.indexOf("**", i + 2)
                if (end > i + 1) {
                    withStyle(SpanStyle(fontWeight = FontWeight.Bold)) { append(text.substring(i + 2, end)) }
                    i = end + 2
                } else {
                    append(text[i]); i += 1
                }
            }
            text[i] == '*' -> {
                val end = text.indexOf('*', i + 1)
                if (end > i) {
                    withStyle(SpanStyle(fontStyle = FontStyle.Italic)) { append(text.substring(i + 1, end)) }
                    i = end + 1
                } else {
                    append(text[i]); i += 1
                }
            }
            text[i] == '`' -> {
                val end = text.indexOf('`', i + 1)
                if (end > i) {
                    withStyle(SpanStyle(fontFamily = FontFamily.Monospace)) { append(text.substring(i + 1, end)) }
                    i = end + 1
                } else {
                    append(text[i]); i += 1
                }
            }
            else -> {
                append(text[i]); i += 1
            }
        }
    }
}
